assignment5/q1.cpp

```diff
@@ -0,0 +1,184 @@
+    for(int i = 0; i < 256; i++) {
+        connections[i] = NULL;
+        in_number[i] = -1;
+    }
+
+    if(dictionary.empty())
+        return;
+
+    if(dictionary.size() == 1) {
+        char word_letter = dictionary.at(0)[0];
+        int length = dictionary.at(0).size();
+        int letter_id = 0;
+        letters.insert(word_letter);
+
+        //if word is made only of one character
+        while((letter_id < length - 1) && word_letter == dictionary.at(0)[letter_id]) {
+            letter_id++;
+        }
+        if(letter_id == length - 1) {
+            connections[word_letter] = &(*new vector<char>);
+            (*connections[(int) word_letter]).push_back(word_letter);
+        }
+
+        return;
+    }
+
+    //make connections
+    for(int word_id = 0; word_id < dictionary.size()-1; word_id++) {
+        string word1 = dictionary[word_id];
+        string word2 = dictionary[word_id + 1];
+        int char_id = 0;
+        while(word1[char_id] == word2[char_id])
+            char_id++;
+
+        if(connections[word1[char_id]] == NULL) {
+            letters.insert(word1[char_id]);
+            in_number[word1[char_id]] = 0;
+            connections[word1[char_id]] = &(*new vector<char>);
+        }
+        if(connections[word2[char_id]] == NULL) {
+            letters.insert(word2[char_id]);
+            in_number[word2[char_id]] = 0;
+            connections[word2[char_id]] = &(*new vector<char>);
+        }
+        //add the directed edge
+        (*connections[word1[char_id]]).push_back(word2[char_id]);
+        letters.erase(word2[char_id]);
+        in_number[word2[char_id]]++;
+
+    }
+
+    return;
+
+}
+
+vector<char> make_alphabet_from_rules(vector<char> *connections[256], set<char> &letters, vector<int> &in_number) {
+
+    int current_id = 0;
+
+    vector<char> alphabet;
+    set<char>::iterator current_letter = letters.begin();
+
+    //if dictionary was empty
+    if(letters.empty()) {
+        cout<<"empty"<<endl;
+        return alphabet;
+    }
+
+    //while there are still letters to process
+    while(!letters.empty()) {
+
+        alphabet.push_back(*current_letter);
+
+        //if dictionary had one word consisting of more than one char
+        if(connections[*current_letter] == NULL){
+            cout<<"cannot make rules"<<endl;
+            return alphabet;
+        }
+
+        vector<char> *vector_for_letter = &(*(connections[*current_letter]));
+
+        //look for all connections from letter
+        for(int vertex = 0; vertex < (*vector_for_letter).size(); vertex++) {
+            char vertex_letter = (*vector_for_letter)[vertex];
+            in_number[vertex_letter]--;
+            if(in_number[vertex_letter] == 0) {
+                letters.insert(vertex_letter);
+            }
+        }
+        //delete letter's connections
+        vector<char> *row = &(*(connections[*current_letter]));
+        letters.erase(*current_letter);
+        (*row).erase((*row).begin(), (*row).end());
```

Comment:
That's what operator `->` is for. `x->y` is equivalent to `(*x).y`.
`row->erase(row->begin(), row->end());`

If you want to clear the vector row, you can use `clear` which is more straightforward:
`row->clear();`